### PR TITLE
Fix MergedContextConfiguration cache key for runtime contexts

### PR DESCRIPTION
**Pull request by sbrannen**
This is a fix for #1262.

A more detailed commit message will be crafted after peer review.

### CODE REVIEW

**Comment by sbrannen** [comment]
Web support is not working with the status quo, so I'll have to take another look.

```
Caused by: java.lang.ClassCastException: class org.springframework.aot.test.AotMergedContextConfiguration cannot be cast to class org.springframework.test.context.web.WebMergedContextConfiguration (org.springframework.aot.test.AotMergedContextConfiguration and org.springframework.test.context.web.WebMergedContextConfiguration are in unnamed module of loader 'app')
	at org.springframework.aot.test.boot.AotSpringBootConfigContextLoader$WebConfigurer.configure(AotSpringBootConfigContextLoader.java:135)
	at org.springframework.aot.test.boot.AotSpringBootConfigContextLoader.loadContext(AotSpringBootConfigContextLoader.java:103)
	at org.springframework.aot.test.boot.AotSpringBootConfigContextLoader.loadContext(AotSpringBootConfigContextLoader.java:50)
	at org.springframework.aot.test.AotCacheAwareContextLoaderDelegate.loadContextInternal(AotCacheAwareContextLoaderDelegate.java:62)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:124)
	... 68 common frames omitted
```

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java [hunk 1]
```diff
@@ -110,7 +116,7 @@ protected ClassName generateTestContext(BootstrapWriterContext writerContext, Su
 		return mainBootstrapClass.getClassName();
 	}
 
-	private MethodSpec.Builder contextLoadersMappingMethod(Map<ClassName, TestContextConfigurationDescriptor> entries) {
+	private MethodSpec.Builder getContextLoadersBuilder(Map<ClassName, TestContextConfigurationDescriptor> entries) {
```
Comment (snicoll) [inline spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:119]: I am using such a pattern for all `MethodSpec.Builder` within the code base and not using the term "builder" in the name of the method was done on purpose as this is a private detail. The same goes for `get` that was removed. 

We can discuss any style change but this has to be done consistently and not as part of this PR.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java [hunk 2]
```diff
@@ -126,6 +132,34 @@ protected ClassName generateTestContext(BootstrapWriterContext writerContext, Su
 				.addModifiers(Modifier.PUBLIC, Modifier.STATIC).addCode(code.build());
 	}
 
+	private MethodSpec.Builder getContextInitializersBuilder(Map<ClassName, TestContextConfigurationDescriptor> entries) {
+		// We're generating a method that looks like the following.
```
Comment (snicoll) [inline spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:136]: Please remove the comments. We now have a method that starts with an example while the existing one does not.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java [hunk 3]
```diff
@@ -126,6 +132,34 @@ protected ClassName generateTestContext(BootstrapWriterContext writerContext, Su
...
+		//     map.put("org.example.Sample2Tests", Sample2TestsContextInitializer.class);
+		//     return map;
+		// }
+
+		ClassName stringTypeName = ClassName.get(String.class);
+		TypeName initializerWildcard = WildcardTypeName.subtypeOf(Object.class);
+		TypeName initializerTypeName = ParameterizedTypeName.get(ClassName.get(ApplicationContextInitializer.class), initializerWildcard);
+		TypeName classWildcard = WildcardTypeName.subtypeOf(initializerTypeName);
+		TypeName classTypeName = ParameterizedTypeName.get(ClassName.get(Class.class), classWildcard);
+		TypeName mapTypeName = ParameterizedTypeName.get(ClassName.get(Map.class), stringTypeName, classTypeName);
```
Comment (snicoll) [inline spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:150]: That's a lot of code to get `Map<String, Class<? extends ApplicationContextInitializer<?>>>`.  The following single statement does the same thing: 

```java
TypeName mapType = ParameterizedTypeName.get(ClassName.get(Map.class),
				ClassName.get(String.class), ParameterizedTypeName.get(ClassName.get(Class.class), WildcardTypeName.subtypeOf(ApplicationContextInitializer.class)));
```

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java [hunk 4]
```diff
@@ -126,6 +132,34 @@ protected ClassName generateTestContext(BootstrapWriterContext writerContext, Su
...
+
+		Builder code = CodeBlock.builder();
+		code.addStatement("$T map = new $T<>()", mapTypeName, HashMap.class);
```
Comment (snicoll) [inline spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:153]: Detail but the field is called `entries` in the other method, why use something else here?

File: spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java [hunk 5]
```diff
@@ -97,22 +98,41 @@ void processWritesContextLoaderMappingAtStandardLocation() {
...
 		ContextBootstrapStructure structure = this.tester.process(SampleApplicationTests.class);
 		assertThat(structure).hasClassDescriptor("com.example.TestContextBootstrapInitializer", (descriptor) -> {
-			assertThat(descriptor.getMethods()).singleElement().satisfies((methodDescriptor) -> {
-				assertThat(methodDescriptor.getName()).isEqualTo("getContextLoaders");
-				assertThat(methodDescriptor.getParameterTypes()).isEmpty();
-			});
 			assertThat(descriptor.getAccess()).isNull();
 			assertThat(descriptor.getFields()).isNull();
+			assertThat(descriptor.getMethods()).hasSize(2);
+			assertThat(descriptor.getMethods()).allSatisfy((methodDescriptor) -> {
```
Comment (snicoll) [inline spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java:130]: I am not fan of the | assertion here. We usually use `anySatisfy` when there is more than one element (there are examples in the codebase).

File: spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java [hunk 6]
```diff
@@ -97,22 +98,41 @@ void processWritesContextLoaderMappingAtStandardLocation() {
...
+
 	@Test
 	void processInvokeTestNativeConfigurationRegistrar() {
 		ContextBootstrapStructure structure = this.tester.process(SampleApplicationTests.class);
 		assertThat(structure).hasResourcePattern("org/springframework/aot/test/samples/app/SampleApplication.class");
 	}
 
 	@Test
-	void processRegisterReflectionForContextLoadersMappingMethod() {
+	void processReflectionRegistrationForMappingMethods() {
```
Comment (snicoll) [inline spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java:124]: Please refrain from doing unrelated changes like this.

File: spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java [hunk 7]
```diff
@@ -38,22 +42,69 @@
...
+		this.aotContextLoader = aotContextLoader;
 	}
 
 	@Override
 	protected ApplicationContext loadContextInternal(MergedContextConfiguration config) throws Exception {
 		SmartContextLoader contextLoader = this.aotContextLoader.getContextLoader(config.getTestClass());
 		if (contextLoader != null) {
-			logger.info("Starting test in AOT mode using " + contextLoader);
-			return contextLoader.loadContext(config);
+			Assert.isInstanceOf(AotMergedContextConfiguration.class, config);
```
Comment (snicoll) [inline spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java:62]: These two lines are unrelated to the task here, isn't it?

File: spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java [hunk 8]
```diff
@@ -38,22 +42,69 @@
 
 	private final AotContextLoader aotContextLoader;
 
+	public AotCacheAwareContextLoaderDelegate() {
+		this(new AotContextLoader());
+	}
+
 	AotCacheAwareContextLoaderDelegate(AotContextLoader aotContextLoader) {
```
Comment (snicoll) [inline spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java:49]: This constructor is no longer called by anything as far as I can see, besides the default constructor. It should be removed.

File: spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java [hunk 9]
```diff
@@ -70,6 +77,24 @@
 		}
 	}
 
+	@SuppressWarnings("unchecked")
```
Comment (snicoll) [inline spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java:80]: Isn't there a case to refactor the existing method to pass the method name and expected type? The cast is happening anyway so I don't see a lot of benefits in having two methods for the reflection business.

File: spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java [hunk 10]
```diff
@@ -79,4 +104,8 @@ boolean isSupportedTestClass(Class<?> testClass) {
 		return this.contextLoaders.containsKey(testClass.getName());
 	}
 
+	Class<? extends ApplicationContextInitializer<?>> getContextInitializerClass(Class<?> testClass) {
```
Comment (snicoll) [inline spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java:107]: I am not against reusing the same component for both purposes (I actually think it's the right call considering the reflection business going on) but I think the class-javadoc should be updated and probably the class renamed as a result.

File: spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+	private final MergedContextConfiguration original;
+
+
+	AotMergedContextConfiguration(Class<?> testClass,
+			Class<? extends ApplicationContextInitializer<?>> contextInitializerClass,
+			MergedContextConfiguration original, AotCacheAwareContextLoaderDelegate cacheAwareContextLoaderDelegate,
+			@Nullable MergedContextConfiguration parent) {
+
+		super(testClass, null, null, Collections.singleton(contextInitializerClass), null, null,
```
Comment (snicoll) [inline spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:54]: `ContextLoader` can't be null according to the contract. We can ignore that if we choose to but I thought it was worth raising.

File: spring-native/src/test/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegateTests.java [hunk 12]
```diff
@@ -45,33 +52,107 @@
...
-		assertThat(delegate.loadContextInternal(createMergedContextConfiguration(SampleTest.class)))
-				.isSameAs(this.applicationContext);
-		verify(this.contextLoader).loadContext(any(MergedContextConfiguration.class));
+	void loadContextWithMatchUsesAotInfrastructure() throws Exception {
+		AotCacheAwareContextLoaderDelegate delegate = createDemoDelegate();
+
+		MergedContextConfiguration mergedConfig = createMergedContextConfiguration(SampleTest.class);
+		assertThat(delegate.loadContext(mergedConfig)).isSameAs(this.applicationContext);
+
+		// loadContext() must actually be invoked with the original MergedContextConfiguration
```
Comment (snicoll) [inline spring-native/src/test/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegateTests.java:69]: It will take me a bit of time to process this but, at a quick glance, I generally feel that a "lot' of explanations like that in a test is the sign that something else should be made more explicit.

File: spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+	/**
+	 * Get the original {@link MergedContextConfiguration} that this
+	 * {@code AotMergedContextConfiguration} was created for.
+	 */
+	MergedContextConfiguration getOriginal() {
+		return this.original;
+	}
+
+	@Override
+	public boolean equals(@Nullable Object other) {
```
Comment (snicoll) [inline spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:69]: I see 0 test hits on this method which seems a little strange.

**Review by snicoll — COMMENTED** [review]
As requested by @sdeleuze, here's a first review. Happy to chat if more details are needed.

**Comment by sdeleuze** [comment]
Please let me know when this PR is ready for final review and merge.

**Comment by sbrannen** [comment]
> Please let me know when this PR is ready for final review and merge.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 1 above)
Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:119]: OK. I've reverted that and renamed the new method to `contextInitializersMappingMethod` for consistency.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 2 above)
Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:136]: I disagree with the rationale.

When JavaPoet is used to generate code, it is not readily clear what is being generated. Thus, the inclusion of the comment is beneficial to anyone reading the code and especially to anyone who has to modify or maintain the code.

Instead of removing this comment, I suggest we add a similar comment to `contextLoadersMappingMethod()` for consistency.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 3 above)
Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:150]: The existing code produces:

```java
Map<String, Class<? extends ApplicationContextInitializer<?>>>
```

The proposed change produces the following which loses the `<?>`:

```java
Map<String, Class<? extends ApplicationContextInitializer>>
```

Although the proposed change is technically a single statement, it's 202 characters long, and I don't think it improves readability or maintainability of the code.

In any case, I'll see about reducing the number of local variables currently in use.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 4 above)
Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:153]: Speaking of "field", I actually was wondering about converting it from a local variable to a `static` field, since we're currently recreating the maps each time one of these methods is invoked (which happens a lot for `getContextInitializers()`).

Had you considered that?

Was there a technical reason to create the map as a local variable instead of as a static field?

Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:153]: With regard to the name of the variable, every map contains entries. So unless the term "entries" has special meaning for the context in which a map is used, I always prefer the generic term `map` unless I think it's necessary to name the map something  more specific.

Since this is (currently) a local variable, I don't see the need for anything more specific than `map`.

With regard to using something different here, I didn't want to change the implementation of the other method, because I know you do not like changes that you do not consider relevant to the current PR.

File: spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java (hunk 5 above)
Comment (sbrannen) [reply spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java:130]: OK. I'll take a look at other usage of `anySatisfy` and see if I can apply that here.

File: spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java (hunk 6 above)
Comment (sbrannen) [reply spring-aot-test/src/test/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessorTests.java:124]: Handled separately in 4787862c319fc609821922102f41e991abb5b1ef

File: spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java (hunk 7 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java:62]: The log statement has been handled separately in 4787862c319fc609821922102f41e991abb5b1ef.

The other line is related to this PR.

File: spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java (hunk 8 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegate.java:49]: Good catch! 👍 

I simply overlooked that.

It's gone now.

File: spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java (hunk 11 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:54]: Yeah, I noticed that as well, but thanks for noticing and mentioning it (b/c I might have overlooked it)!

A non-null `ContextLoader` is in fact required by `DefaultCacheAwareContextLoaderDelegate.loadContextInternal(...)`, but we override that method and ignore the `ContextLoader` in the supplied `MergedContextConfiguration` since we are using the `AotSpringBootConfigContextLoader`.

So I think it's safe to ignore this for our use case.

File: spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java (hunk 13 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:69]: `equals()` and `hashCode()` are used by the `LruCache` in `DefaultContextCache`.

So if you don't see any hits, that means that the context cache is not being exercised.

File: spring-native/src/test/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegateTests.java (hunk 12 above)
Comment (sbrannen) [reply spring-native/src/test/java/org/springframework/aot/test/AotCacheAwareContextLoaderDelegateTests.java:69]: Fair enough.

I'll move the comments to actual Javadoc in the production code.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 2 above)
Comment (snicoll) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:136]: Thanks but I still don't think this is a good idea, certainly for the reason you've shared. 

The tests are meant to show what is generated rather than a comment that is going to be outdated. If you need a visual representation of what the method will look like concretely, write a test for it that asserts the content. That's what do everywhere else.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 3 above)
Comment (snicoll) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:150]: Not going to comment any further on the 202 remark besides the fact that loosing variable assignment was the goal here. Sorry I missed the extra wildcard.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 4 above)
Comment (snicoll) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:153]: > With regard to using something different here, I didn't want to change the implementation of the other method, because I know you do not like changes that you do not consider relevant to the current PR.

That's exactly the point that I am trying to raise for several PRs in a row now. If you disagree about something, that's fine to discuss it and then come to a conclusion but changing it in your own code just brings inconsistency and I don't like that.

File: spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java (hunk 13 above)
Comment (snicoll) [reply spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:69]: > So if you don't see any hits, that means that the context cache is not being exercised.

Exactly. The context cache is not exercised while running all tests in that package concerns me. Shouldn't it be used?

File: spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java (hunk 9 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java:80]: I was actually considering doing that but hadn't gotten to it.

It's done now (locally).

File: spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java (hunk 10 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotContextLoader.java:107]: I agree. 

I'll come up with a new name and update the documentation.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 2 above)
Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:136]: OK. I'll remove the comment.

I already have a test for it that asserts the content.

File: spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java (hunk 4 above)
Comment (sbrannen) [reply spring-aot-test/src/main/java/org/springframework/aot/test/context/bootstrap/generator/TestContextAotProcessor.java:153]: > Speaking of "field", I actually was wondering about converting it from a local variable to a static field, since we're currently recreating the maps each time one of these methods is invoked (which happens a lot for getContextInitializers()).

You can disregard that. I think that used to be the case, but it no longer is. The methods that create the maps should only be invoked twice now: once for `AotCacheAwareContextLoaderDelegate` and once for `AotDependencyInjectionTestExecutionListener`.

File: spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java (hunk 13 above)
Comment (sbrannen) [reply spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:69]: Ahhh... I mistakenly thought you were questioning the need for overriding `equals()`.

> Shouldn't it be used?

It's used in the sample projects as functional tests, but I'll see if I can introduce an integration test that exercises those methods via the context cache.

Comment (snicoll) [reply spring-native/src/main/java/org/springframework/aot/test/AotMergedContextConfiguration.java:69]: > Ahhh... I mistakenly thought you were questioning the need for overriding equals().

On the contrary. Implementing `equals` and `hashCode` is at the hear of the PR so not having tests for those felt strange. Simple unit tests should be enough though, no need for having to deal with the context cache.

**Comment by sbrannen** [comment]
This PR is ready for a final review.

The only remaining task is an update to the Javadoc (see https://github.com/spring-projects-experimental/spring-native/pull/1341#discussion_r763187815), but that can be done at later date if necessary.

**Comment by snicoll** [comment]
Thanks Sam, this looks good.
